javagen/src/main/java/com/azure/autorest/template/StreamSerializationModelTemplate.java

```diff
@@ -369,17 +463,35 @@ private static void serializeJsonContainerProperty(JavaBlock methodBlock, String
                 methodBlock.line(lambdaWriterName + ".writeFieldName(\"" + serializedName + "\");");
                 methodBlock.line(ClientModelUtil.CORE_TO_CODEGEN_BRIDGE_UTILS_CLASS_NAME + ".responseErrorToJson(" + lambdaWriterName + ", " + propertyValueGetter + ");");
             } else if (valueSerializationMethod != null) {
-                methodBlock.line(valueSerializationMethod);
+                if (isJsonMergePatch) {
+                    if (elementType.isNullable()) {
+                        methodBlock.block("", codeBlock -> {
+                            codeBlock.ifBlock(elementName + "!=null", ifBlock -> {
+                                if (elementType instanceof ClassType && ((ClassType) elementType).isSwaggerType()) {
+                                    codeBlock.line(elementName + ".serializeAsJsonMergePatch(true);");
+                                }
+                                ifBlock.line(valueSerializationMethod + ";");
+                                if (elementType instanceof ClassType && ((ClassType) elementType).isSwaggerType()) {
+                                    codeBlock.line(elementName + ".serializeAsJsonMergePatch(false);");
+                                }
+                            }).elseBlock(elseBlock -> elseBlock.line(lambdaWriterName + ".writeNull();"));
+                        });
+                    } else {
+                        methodBlock.line(valueSerializationMethod);
+                    }
```

Comment:
Is this for the Map?

If yes, all Map element in json-merge-patch should be treated as nullable. So I don't think we should test that (as isNullable here is the state of property in shared model).

Even if element is primitive integer, it should still test for null, as the type is `Map<String, Integer>` and user can still set it to null to remove a key-value.

Reply:
I add it mainly for the Map case. But this is for serializing container case, e.g. `Map`, `Array`, `List`. I'm thinking if it is `int[]` array, the parameter can be `int`. So I added this for safety reason.

Reply:
So far, codegen don't have `int[]`. `ArrayType` is only for `byte[]`.

Reply:
Or you may target Map only?

I don't really sure what's
`List['a', 'b', null]` means for json-merge-patch. That probably not valid.

Reply:
ok, if that's the case, I might simply remove the null check.

Reply:
I mean Map should have it in any case (this is a json-merge-patch to clean a value -- Map is just a JSON object, same as your Class model). List/Array may not need any special handling in json-merge-patch (I think `null` here does not have special meaning, but safer to double check RFC).

Reply:
I see, we may only need to do special handling for json-mergpatch+Map case.

Reply:
Updated code, add a test case for array. If we remove nullable check here, when we set one element of array to `null`, it will not be serialized.
```
@Test
    public void testSerializationForArrayProperty() throws JsonProcessingException {
        Resource resource = new Resource(new HashMap<>());
        resource.setArray(Arrays.asList(new InnerModel("value1"), new InnerModel("value2")));
        JsonMergePatchHelper.getResourceAccessor().prepareModelForJsonMergePatch(resource, true);
        resource.getMap().put("key", null);
        resource.getArray().set(0, null);
        String json = BinaryData.fromObject(resource).toString(); // {"map":{"key":null},"array":[{"name":"value2"}]}
        JsonNode node = OBJECT_MAPPER.readTree(json);
        Assertions.assertEquals(JsonNodeType.NULL, node.get("map").get("key").getNodeType());
        Assertions.assertEquals(1, node.get("array").size());
        Assertions.assertEquals("value2", node.get("array").get(0).get("name").asText());
    }
```